File: Python/373.Find K Pairs with Smallest Sums.py
#给出两个升序数组和一个数k，求和最小的k个数对，两个数分别来自两个数组。

#在向堆中添加新元素时，只考虑i, j + 1；（当j == 0时额外考虑i + 1, 0）
import heapq
class Solution(object):
    def kSmallestPairs(self, nums1, nums2, k):
        queue = []
        def push(i, j):
            if i < len(nums1) and j < len(nums2):
                heapq.heappush(queue, [nums1[i] + nums2[j], i, j])
        push(0, 0)
        pairs = []
        while queue and len(pairs) < k:
            _, i, j = heapq.heappop(queue)
            pairs.append([nums1[i], nums2[j]])
            push(i, j + 1)
            if j == 0:
                push(i + 1, 0)
        return pairs



s = Solution()
print(s.kSmallestPairs([1,7,11],[2,4,9],8))

File: C++/Heap/373.Find K Pairs with Smallest Sums.cpp
/*
给出两个有序数组和数k，从两个数组中各拿一个组成数对，
问和最小的k个数对是哪些?
 */
/*
思路：使用一个indice数组，indice[i]表示当前状态下，
	  nums1数组中的第i个数已经加入结果中的pair个数。
	  每次循环加入一个pair，对应的下标加1。
	  在寻找每一个和最小的pair过程中，对nums1中每个数nums1[i]循环，
	  配对的pair分别就是nums2[indice[i]]。
 */
#include <iostream>
#include <vector>
#include <utility>
using namespace std;

class Solution {
public:
    vector<pair<int, int>> kSmallestPairs(vector<int>& nums1, vector<int>& nums2, int k) {
        vector<pair<int, int>> res;
        int m = nums1.size() , n = nums2.size();
        k = min(k, m * n);
        vector<int> indice(m, 0);
        while(k-- > 0)
        {
            int tmp_index = 0;
            long tmp_sum = LONG_MAX;
            for(int i = 0; i < m; i++)
            {
                if(indice[i] < n && tmp_sum >= nums1[i] + nums2[indice[i]])
                {
                    tmp_index = i;
                    tmp_sum = nums1[i] + nums2[indice[i]];
                }
            }
            res.push_back(make_pair(nums1[tmp_index], nums2[indice[tmp_index]]));
            indice[tmp_index]++;
        }
        return res;
    }
};
int main()
{
	Solution s;
	vector<int> nums1{1,7,11},nums2{2,4,6};
	vector<pair<int, int>> out =  s.kSmallestPairs(nums1,nums2,3);
	for(auto p:out)
		cout<<p.first<<" "<<p.second<<endl;
}

